dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalTransformer.kt

```diff
@@ -0,0 +1,97 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.tools.dokka.plugin
+
+import io.spine.annotation.Internal
+import org.jetbrains.dokka.base.transformers.documentables.SuppressedByConditionDocumentableFilterTransformer
+import org.jetbrains.dokka.model.Annotations
+import org.jetbrains.dokka.model.Annotations.Annotation
+import org.jetbrains.dokka.model.Documentable
+import org.jetbrains.dokka.model.properties.WithExtraProperties
+import org.jetbrains.dokka.plugability.DokkaContext
+
+/**
+ * Excludes everything annotated with [Internal] from the resulting documentation.
+ *
+ * Works for packages, types, fields and methods.
+ */
+public class ExcludeInternalTransformer(dokkaContext: DokkaContext) :
+    SuppressedByConditionDocumentableFilterTransformer(dokkaContext) {
+
+    /**
+     * Tells whether the passed [Documentable] should be excluded from the documentation.
+     *
+     * Returns `true` if the passed piece is annotated with the [Internal] annotation. Otherwise,
+     * returns `false`.
+     */
+    override fun shouldBeSuppressed(d: Documentable): Boolean {
+        /**
+         * [WithExtraProperties] is a Dokka-interface that adds a map-field for extra properties.
+         * The information about applied annotations is present in this map.
+         *
+         * Not every [Documentable] subclass implements this interface.
+         */
+        return (d is WithExtraProperties<*>) && hasInternalAnnotation(d)
+    }
+
+    private fun hasInternalAnnotation(annotated: WithExtraProperties<*>): Boolean =
+        annotated.annotations().any(InternalAnnotationCheck::test)
+
+    /**
+     * Extracts annotations from a [WithExtraProperties] object.
+     *
+     * Information about annotations is found in the map-field `extra`, added when implementing
+     * [WithExtraProperties].
+     */
+    private fun WithExtraProperties<*>.annotations(): List<Annotation> {
+        /**
+         * All values with the type of [Annotations], a container-class for all applied annotations,
+         * are extracted from the map-field. Then plain lists of [Annotation] are extracted from
+         * every [Annotations] object and merged into the result list.
+         */
+        return this.extra.allOfType<Annotations>().flatMap {
+            it.directAnnotations.values.flatten()
+        }
+    }
+
+    /**
+     * Provides the method to check if an [Annotation] object represents [Internal] annotation.
+     */
+    private object InternalAnnotationCheck {
+        private val c = Internal::class.java
+
+        /**
+         * Checks if provided object represents [Internal] annotation.
+         *
+         * Compares the package and the class extracted from the object with the package
+         * and the class of the [Internal] annotation.
```

Comment:
Judging on this description, I have doubts on whether the whole thing would work for the packages which are annotated with `@Internal` in their `package-info.java`. Let's check that out of scope of this PR (please create a respective GitHub issue in this repository).

Reply:
I don't see why that should not work. Dokka collects annotations applied to the piece of code(whether it is a package, a type, or a method). We then extract this information as `List<Annotation>` and go one by one through the list checking if the current annotation represents `@Internal`. The check algorithm is described in the comments you mentioned.

Reply:
I was wrong. By analyzing the source code of Dokka, I found out that they indeed do not check packages with `shouldBeSupressed`, which we overrode a few lines above. They immediately go to processing types, methods, etc. If it was not the case, this check would work as it is concept-agnostic and compares provided annotation with `@Internal`.